File: quest-sys/src/decoherence.cpp
//
// Created by Erich Essmann on 12/03/2025.
//
#include "bindings.h"
#include "helper.h"

namespace quest_sys {
void mixDephasing(Qureg& qureg, int qubit, qreal prob) {
  ::mixDephasing(qureg, qubit, prob);
}

void mixTwoQubitDephasing(Qureg& qureg, int qubit1, int qubit2, qreal prob) {
  ::mixTwoQubitDephasing(qureg, qubit1, qubit2, prob);
}

void mixDepolarising(Qureg& qureg, int qubit, qreal prob) {
  ::mixDepolarising(qureg, qubit, prob);
}

void mixTwoQubitDepolarising(Qureg& qureg, int qubit1, int qubit2, qreal prob) {
  ::mixTwoQubitDepolarising(qureg, qubit1, qubit2, prob);
}

void mixDamping(Qureg& qureg, int qubit, qreal prob) {
  ::mixDamping(qureg, qubit, prob);
}

void mixPaulis(Qureg& qureg, int qubit, qreal probX, qreal probY, qreal probZ) {
  ::mixPaulis(qureg, qubit, probX, probY, probZ);
}

void mixQureg(Qureg& qureg, Qureg& other, qreal prob) {
  ::mixQureg(qureg, other, prob);
}

void mixKrausMap(Qureg& qureg,
                 rust::Slice<const int> qubits,
                 const KrausMap& map) {
  ::mixKrausMap(qureg, quest_helper::slice_to_ptr(qubits),
                static_cast<int>(qubits.size()), map);
}
}  // namespace quest_sys

File: quest-sys/src/decoherence.rs
#[cxx::bridge]
pub mod ffi {
    // Types
    unsafe extern "C++" {
        // Opaque QuEST types
        type Qureg;
        type KrausMap;

        // Common type
        type Quest_Complex = crate::types::ffi::Quest_Complex;
    }

    // Decoherence
    #[namespace = "quest_sys"]
    unsafe extern "C++" {
        include!("decoherence.hpp");
        // Decoherence functions
        fn mixDephasing(qureg: Pin<&mut Qureg>, qubit: i32, prob: f64);
        fn mixTwoQubitDephasing(qureg: Pin<&mut Qureg>, qubit1: i32, qubit2: i32, prob: f64);
        fn mixDepolarising(qureg: Pin<&mut Qureg>, qubit: i32, prob: f64);
        fn mixTwoQubitDepolarising(qureg: Pin<&mut Qureg>, qubit1: i32, qubit2: i32, prob: f64);
        fn mixDamping(qureg: Pin<&mut Qureg>, qubit: i32, prob: f64);
        fn mixPaulis(qureg: Pin<&mut Qureg>, qubit: i32, probX: f64, probY: f64, probZ: f64);
        fn mixQureg(qureg: Pin<&mut Qureg>, other: Pin<&mut Qureg>, prob: f64);
        fn mixKrausMap(qureg: Pin<&mut Qureg>, qubits: &[i32], map: &KrausMap);
    }
}
